### ISSUE TITLE
keyboard [enter] , click OK.png not working - see working example from Marcelo

### ISSUE THREAD

**Issue by maheswarGR** [kelaberetiv/TagUI#1171]
@kensoh   Using windows tagui 6 
Need to click OK or cancel button in alert window. Tried multiple ways
keyboard [enter]  
click OK.png 
vision click("OK.png")
not working 

Also tried taking screenshot using sikuli and placed the image in C:\tagui\flows\samples.. not working

I have placed images in C:\tagui\flows\samples

<img width="54" alt="OK" src="https://user-images.githubusercontent.com/91525476/151301049-fe573b9e-b11f-4d7d-86ba-cf3192d04c82.PNG">

![image](https://user-images.githubusercontent.com/91525476/151300958-4746866e-93f5-4b59-86c2-ee6110b9c1e5.png)

**Comment by kensoh** [comment]
Hi @maheswarGR the image has to be in the same folder as your .tag file, if you are doing `click OK.png`.

Can you try changing your Windows display zoom % to the recommended % or 100% to see if that helps?

Also, can you share the log file in tagui\src\tagui.sikuli folder to see if there are clues there?

**Comment by kensoh** [comment]
CC @ruthtxh

**Comment by kensoh** [comment]
Lastly, make sure you have installed Java (for eg Open Corretto), at least 64-bit version 8 is needed
See more here and the link to the Corretto installer - https://tagui.readthedocs.io/en/latest/setup.html

**Comment by maheswarGR** [comment]
yes image.png and tag files are in same path C:\tagui\flows\samples
 
 log file tagui.sikuli  is empty 
 How to chnage windows display zoom ?

**Comment by kensoh** [comment]
Have you tried installing Corretto from above link?

For changing Windows display zoom, you can google to find tutorial or video to show the steps.

**Comment by kensoh** [comment]
You can also join our weekly Zoom call in 40 minutes time to look into this

**Comment by kensoh** [comment]
Here is the URL - https://github.com/kelaberetiv/TagUI/issues/914

**Comment by maheswarGR** [comment]
yes windows zoom is previously 100% and i tried changing to recommended % also ..didnt work

I am using java version

java version "1.8.0"
Java(TM) SE Runtime Environment (build 1.8.0-b132)
Java HotSpot(TM) 64-Bit Server VM (build 25.0-b70, mixed mode)

will this support tagui vision click?

**Comment by kensoh** [comment]
Yes this version is ok. Can you share the logs here in tagui\src\tagui.sikuli folder to see if there are clues there?

**Comment by maheswarGR** [comment]
there are many files in it
expect 1st file , all are empty
![image](https://user-images.githubusercontent.com/91525476/151325376-3d068f9b-8120-47c1-88ab-129a7ac7f1ba.png)

**Comment by kensoh** [comment]
Oh I see, I know why now. The reason is because until user manually clicks the OK button, the control does not return back to TagUI. The solution is using visual automation to click on the button that triggers this popup, then clicking OK. For eg

```
click delete.png
click ok.png
```

Please try something like above and tell me what do you see in the logs, thanks!

**Comment by maheswarGR** [comment]
I tried .  Until user manually clicks the OK button, the control does not return back to TagUI...
"click 1643348149398.png" in tag file is not read.
the control is not coming back to the tagui from chrome when the alert is displayed

I have attached log files for your reference..

My tag file :

http://3.132.253.47/dmapim/publisher/login.html?publisherName=coe

wait 15s
type //*[@id="username_view"] as admin
type //*[@id="password"] as Thbs1234!
click //*[@id="container"]/div/center/form/div/div[4]/div/button
wait 20s

click //*[@id="menuAccordion"]/div[1]/a/span
click //td[text()='RPA-PoC-Blueprism']//following::td[@id='delete']/i

click 1643348149398.png


[logFiles.zip](https://github.com/kelaberetiv/TagUI/files/7956185/logFiles.zip)

**Comment by kensoh** [comment]
Hi @maheswarGR , as mentioned above, use `click delete.png` to use visual automation instead of `click //td[text()='RPA-PoC-Blueprism']//following::td[@id='delete']/i` because `click //td[text()='RPA-PoC-Blueprism']//following::td[@id='delete']/i` will hang TagUI process and not return control until user clicks OK.

**Comment by maheswarGR** [comment]
yes i tried .but facing a problem there. 
delete.png - ![delete](https://user-images.githubusercontent.com/91525476/151503429-758f6755-456c-443c-9af5-bc6c0e7ea9cd.png)

but it is clicking on close browser . How can i avoid this ?
![image](https://user-images.githubusercontent.com/91525476/151503676-427b425b-0656-457d-a78a-be6ade4c5d75.png)

My page :
 My page has multiple delete images. Need to click on RPA-PoC-Blueprism delete image.
![image](https://user-images.githubusercontent.com/91525476/151504245-ca9e7c07-5dee-4164-818d-7a93a13f9fc4.png)

Modified tag file :
............
wait 20s
click //*[@id="menuAccordion"]/div[1]/a/span

click delete.png
click ok.png

**Comment by kensoh** [comment]
Oh yes it looks very similar. Is the delete to be click on always the first row record? If the row to click on is always different it will be hard. Or is it always at the same (x,y) position on the webpage? Thinking of other ways to click on the correct X.

**Comment by marcelocecin** [comment]
hello guys, another way would be to get the position of the icon next to it (pencil) and add the width between them to send the mouse click command

**Comment by maheswarGR** [comment]
HI @marcelocecin @kensoh ..
 It is  not same (x,y) position on the webpage everytime ..requirement is to delete with respect to server name (i.e.,RPA-PoC-Blueprism)..so tried using xpath ..
 
 Following script worked out:
//click //td[text()='RPA-PoC-Blueprism']//following::td[@id='delete']/i

click (1855,615)
wait 5s
hover cancel.png
click ok.png

1. Can you kindly brief reason for why control is not returning (hangs) to TAGUI process after 
           > click td[text()='RPA-PoC-Blueprism']//following::td[@id='delete']/i  
           
2. Is there any approach to make click with respect to server name (i.e.,RPA-PoC-Blueprism) or alternate way of using xpath without hanging TAGUI process?

**Comment by kensoh** [comment]
1. The reason is this https://github.com/kelaberetiv/TagUI/issues/1171#issuecomment-1023941567 - Chrome by design does not return control to TagUI when there is a popup dialog.

2. Maybe you can try `click Autostub using ocr` by using OCR to search for a portion of the server name

**Comment by marcelocecin** [comment]
`click //td[contains(text(),'RPA PoC with Blueprism')]/../td[9]`

**Comment by maheswarGR** [comment]
sikuli screnshot - how to include restricted region in tag script

@kensoh hi, just took a screenshot of an icon and selected restricted region using sikuli ..After saving it has created an png and python source file ..

how can i include the restricted region in tag script commands i.e., in click delete.png

**Comment by marcelocecin** [comment]
if you want help about python, the right thing is to create an issue in the repository https://github.com/tebelorg/RPA-Python

**Comment by kensoh** [comment]
Oh nice, thanks Marcelo for sharing this!

Hi @maheswarGR , for your question 1 and 3 they seems to be the same question, I've replied here at the new issue you've created - https://github.com/kelaberetiv/TagUI/issues/1174

For the second question on hover, you can try double-checking in Chrome web inspector to see if that XPath really exists. You can also check your Chrome is at 100% zoom. Otherwise TagUI will hover at the wrong position.

**Comment by maheswarGR** [comment]
> if you want help about python, the right thing is to create an issue in the repository https://github.com/tebelorg/RPA-Python I tested here and it worked, follow the script:
> 
> ```
> http://3.132.253.47/dmapim/publisher/login.html?publisherName=coe
> wait 5
> if present('//*[@id="username_view"]')
>   type //*[@id="username_view"] as admin
>   type //*[@id="password"] as Thbs1234![enter]
>   wait 5
>   if present('Backends')
>     click Backends
>     wait 5
>     click (//td[contains(text(),"RPA PoC with Blueprism")]/../td)[9]
> ```
> 
> <img alt="" width="1362" src="https://user-images.githubusercontent.com/2955762/152134980-eeca3591-1d75-411d-a019-92164d5686c2.png">

@marcelocecin yes that worked ..we faced issue only after that alert rises, couldn't click on ok or cancel button..it hanged ..so we done last click with image and followed to click ok.png

**Comment by maheswarGR** [comment]
> Oh nice, thanks Marcelo for sharing this!
> 
> Hi @maheswarGR , for your question 1 and 3 they seems to be the same question, I've replied here at the new issue you've created - #1174
> 
> For the second question on hover, you can try double-checking in Chrome web inspector to see if that XPath really exists. You can also check your Chrome is at 100% zoom. Otherwise TagUI will hover at the wrong position.

@kensoh xpath exists and chrome is at 100% zoom .. i tried click with xpath it works..hover xpath doesn't work ..
    I tried hover xpath for various existing xpaths - nothing works ..but same click xpaths works
    
![image](https://user-images.githubusercontent.com/91525476/152279122-3d658f8b-2a45-46ca-9590-5479cbe6b78a.png)

**Comment by kensoh** [comment]
I see.. In that case, it sounds like the web element does not react to the hover event from TagUI.

You can try using visual automation to move the mouse to that location to hover. Make sure to click on Chrome browser first to bring it to the foreground and let it be in focus.

**Comment by maheswarGR** [comment]
yes .. below one works .
hover (1800,200)
hover delete.png 

BUT why hover xpath doesn't work ?

**Comment by marcelocecin** [comment]
you can use [keyboard commands](https://tagui.readthedocs.io/en/latest/reference.html#keyboard): enter to confirm the operation or esc to cancel
```
keyboard [enter]
```

**Comment by kensoh** [comment]
Adding, if zoom is already at 100%, and hover doesn't work, it means that the web element can't receive the hover event sent from TagUI. So you'll have to find other means to get to the outcome you want.

**Comment by ruthtxh** [comment]
Closing this issue due to inactivity, feel free to reopen.

**Comment by Turbolento** [comment]
@kensoh 
 > Adding, if zoom is already at 100%, and hover doesn't work, it means that the web element can't receive the hover event sent from TagUI. So you'll have to find other means to get to the outcome you want.

Hi, bro, I have the same problem — hover xpath doesn't work , but click xpath works.
About " find other means", I don't quite understand it, I look forward to your more specific solution.

**Comment by ruthtxh** [comment]
Hi @Turbolento perhaps you can try using coordinates or visual automation (screenshot image) and see if hover works for these?

```
hover (100,100)
```

```
hover image.png
```
where you can save the image.png to the same folder as your flow file

**Comment by kensoh** [comment]
You can try to give a simple replication example of TagUI script why hover does not work, then @ruthtxh can have a look. So far, users have not feed back that hover step does not work. It may be something specific to the website you are automating.

**Comment by Turbolento** [comment]
> You can try to give a simple replication example of TagUI script why hover does not work, then @ruthtxh can have a look. So far, users have not feed back that hover step does not work. It may be something specific to the website you are automating.

Thank you for your very nice advice.
This is a example rpa_for_python script:

      import rpa as r
      r.init(visual_automation=True)
      r.url("https://www.google.com/doodles")
      r.hover("//a[@id='logo']")   # not work
      r.hover("1.png")      # work well   
      r.hover(100,50)       # work well

The same thing happens with the following Tagui script:

      tagui live
      https://www.google.com/doodles
      hover //a[@id='logo']
      hover 1.png
      hover (100,200)

![1](https://user-images.githubusercontent.com/30725265/176076114-52e4734c-6165-42a3-9d1a-2e60f9c6651a.gif)

**Comment by ruthtxh** [comment]
Hi @Turbolento 

For hover step with XPath, chrome controls the element event from the backend and upon hover the element is highlighted without the mouse moving to it visibly. For this logo link, upon hover there is no visible change. You can try this on other elements that might have visible change on hover to try this out. On the other hand, for hovering with coordinates and visual automation, the mouse event is directly controlled to move to the location.

**Comment by Turbolento** [comment]
So I can't get the element's coordinates exactly by `hover xpath & echo `mouse_xy()``, right?

**Comment by kensoh** [comment]
There are some internal TagUI functions that could be used to achieve this. But it is not the most commonly used, so there are not dedicated functions to do this. See below working example. chrome.mouse.getXY() is the internal function to get XY coordinates of an element on a webpage, xps666() is to convert a XPath string into a form used by TagUI underlying engine. If it is a CSS selector, xps666() is not required. Fyi @ruthtxh 

```
https://google.com
position = chrome.mouse.getXY(xps666('//*[@name="q"]'))
echo `position.x`,`position.y`
```

**Comment by kensoh** [comment]
See example below. It is not straightforward, because this is not a common use case, but doable from RPA for Python -

```py
r.url('https://google.com')
r.send('xy = chrome.mouse.getXY(xps666(\'//*[@name="q"]\'))')
r.send('dump xy.x to x.txt')
r.send('dump xy.y to y.txt')
x = int(r.load('x.txt'))
y = int(r.load('y.txt'))
print(x,y)
```

PS - make sure to escape double and single quotes as needed in send() so that it is valid Python string and not error
